Web Inspector: Make frontend engineering settings available behind a debug runtime flag

patrickangle opened the pull request:
#### 2326b1146886546de0d44f6aa85415b2e1257b8a
<pre>
Web Inspector: Make frontend engineering settings available behind a debug runtime flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=244819">https://bugs.webkit.org/show_bug.cgi?id=244819</a>
rdar://98957031

Reviewed by Devin Rousso.

Make it slightly easier to use the engineering settings by allowing platforms to define various additional circumstances
under which engineering settings are available.

This is not done as additional functionality behind `DeveloperExtrasEnabled`, as that setting is currently used in some
applications to enable the &quot;Inspect Element&quot; context menu item, but these settings are not generally useful for the
users of those applications.

&quot;Debug&quot; settings and functionality remain limited to local engineering builds.

* Source/WTF/Scripts/Preferences/WebPreferencesDebug.yaml:
- Add a runtime preference to control engineering setting availability.

* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::engineeringSettingsAllowed):
* Source/WebCore/inspector/InspectorFrontendHost.h:
* Source/WebCore/inspector/InspectorFrontendHost.idl:

* Source/WebInspectorUI/UserInterface/Base/Main.js:
* Source/WebInspectorUI/UserInterface/Test/Test.js:
* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.prototype.initializeTarget):
* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.async initializeTarget):
- Add helper to determine if engineering settings should be available based on both the preference and the actual build.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
(WI.EngineeringSetting.prototype.get value):
(WI.EngineeringSetting.prototype.set value):
(WI.EngineeringSetting):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype.initialLayout):
(WI.SettingsTabContentView.prototype._createEngineeringSettingsView):
(WI.SettingsTabContentView.prototype._createDebugSettingsView):
* Source/WebInspectorUI/UserInterface/Debug/Bootstrap.js:
(updateMockWebExtensionTab):
(WI.runBootstrapOperations):
- Move the mock inspector extension tab setting to the Debug section, since it relies on Debug/Bootstrap.js being in the
build.

Canonical link: <a href="https://commits.webkit.org/254676@main">https://commits.webkit.org/254676@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d07e8dd5a63c1cb189ceda20c0dbe999b4e5899e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89797 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34346 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99123 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155940 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93804 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32837 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28285 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82148 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93473 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95444 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26093 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76626 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26025 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80813 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69023 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81452 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30589 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14897 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76319 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30334 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15834 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26440 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3284 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33788 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38786 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/sandbox-toggle-in-inactive-document-crash.html (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/78905 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32500 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34922 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17286 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/InspectorFrontendHost.cpp:
```diff
@@ -636,6 +640,17 @@ void InspectorFrontendHost::setAllowsInspectingInspector(bool allow)
+    return allowsInternalSecurityPolicies && m_frontendPage->settings().webInspectorEngineeringSettingsAllowed();
+#else
+    return false;
+#endif
```
NIT: I'd maybe reorganize this to avoid the `#else` and whatnot
```
    if (!m_frontendPage->settings().webInspectorEngineeringSettingsAllowed())
        return false;

#if PLATFORM(COCOA)
    static bool allowsInternalSecurityPolicies = os_variant_allows_internal_security_policies("com.apple.WebKit");
    if (!allowsInternalSecurityPolicies)
        return false;
#endif

    return true;
```
i.e. other platforms might also want the use of `WebInspectorEngineeringSettingsAllowed` :)

patrickangle replied:
Good call.

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -3580,6 +3580,10 @@ WI.reset = async function()
 WI.isEngineeringBuild = false;
 
+WI.engineeringSettingsAllowed = function() {
+    return WI.isEngineeringBuild || InspectorFrontendHost.engineeringSettingsAllowed();
```
Do we even need `WI.isEngineeringBuild` anymore?  Or is this just a convenience to allow engineers to skip changing that preference for local builds?

Also, should we maybe cache the result of `InspectorFrontendHost.engineeringSettingsAllowed()` somewhere?

patrickangle replied:
Depending on a platform's decision in `InspectorFrontendHost.engineeringSettingsAllows()`, `WI.isEngineeringBuild` can still be useful for open-source contributors to access the functionality.

Caching the result of `InspectorFrontendHost.engineeringSettingsAllows()` makes sense.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -447,7 +448,7 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
     _createEngineeringSettingsView()
     {
-        // These settings are only ever shown in engineering builds, so the strings are unlocalized.
+        // These settings are only ever shown in when engineering tools are enabled or in engineering builds, so the strings are unlocalized.
```
oops "in when"

patrickangle committed:
Web Inspector: Make frontend engineering settings available behind a debug runtime flag
https://bugs.webkit.org/show_bug.cgi?id=244819
rdar://98957031

Reviewed by Devin Rousso.

Make it slightly easier to use the engineering settings by allowing platforms to define various additional circumstances
under which engineering settings are available.

This is not done as additional functionality behind `DeveloperExtrasEnabled`, as that setting is currently used in some
applications to enable the "Inspect Element" context menu item, but these settings are not generally useful for the
users of those applications.

"Debug" settings and functionality remain limited to local engineering builds.

* Source/WTF/Scripts/Preferences/WebPreferencesDebug.yaml:
- Add a runtime preference to control engineering setting availability.

* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::engineeringSettingsAllowed):
* Source/WebCore/inspector/InspectorFrontendHost.h:
* Source/WebCore/inspector/InspectorFrontendHost.idl:

* Source/WebInspectorUI/UserInterface/Base/Main.js:
* Source/WebInspectorUI/UserInterface/Test/Test.js:
* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.prototype.initializeTarget):
* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.async initializeTarget):
- Add helper to determine if engineering settings should be available based on both the preference and the actual build.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
(WI.EngineeringSetting.prototype.get value):
(WI.EngineeringSetting.prototype.set value):
(WI.EngineeringSetting):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype.initialLayout):
(WI.SettingsTabContentView.prototype._createEngineeringSettingsView):
(WI.SettingsTabContentView.prototype._createDebugSettingsView):
* Source/WebInspectorUI/UserInterface/Debug/Bootstrap.js:
(updateMockWebExtensionTab):
(WI.runBootstrapOperations):
- Move the mock inspector extension tab setting to the Debug section, since it relies on Debug/Bootstrap.js being in the
build.

Canonical link: https://commits.webkit.org/254676@main
